lib/reline/ansi.rb

```diff
@@ -7,6 +7,10 @@ class Reline::ANSI
   CAPNAME_KEY_BINDINGS = {
     'khome' => :ed_move_to_beg,
     'kend'  => :ed_move_to_end,
+    'kich1' => :ed_ignore,
+    'kdch1' => :key_delete,
```

Comment:
I was looking at the exact same issue, and I ended up using `:em_delete` for this. `:key_delete` also moves the cursor to the left.

line | type | `key_delete` | `em_delete`
-|-|-|-
| `abc\|` | <kbd>DEL</kbd> | `abc\|` | same
| `ab\|c` | <kbd>DEL</kbd> | `a\|b` | `ab\|`
| `ab\|c` | <kbd>DEL</kbd> <kbd>DEL</kbd> | `\|a` | `ab\|`

Reply:
I admit I kinda like the behavior of `:em_delete` better, but there are issues with mapping to that, and I think what you're really after is a desire to change the behavior of `key_delete`.

For starters, note that the `em_` methods are used and designed to work in emacs mode, not vi insert or vi command mode.  So even if it worked without any quirks, it wouldn't make sense to map to an emacs method, just like it wouldn't make sense to map to a vi-specific `vi_xxx` method.

And it has quirks / problems.  Mapping to `:em_delete` breaks vi command mode in multiple ways.  For starters, the delete key by design is ignored in vi command mode (`key_delete` calls `ed_delete_next_char`, **except** in vi command mode).

By mapping directly to `ed_delete_next_char`, or by mapping to `em_delete`, it breaks this design choice and causes the delete key to work even in command mode, when by design it should not do anything.

(One could argue for wanting the delete key to work in vi command mode, and I would actually agree with that, but if that is wanted it should be a purposeful change to the `key_delete` method).

Now here's where it gets really quirky.  Calling `em_delete` in vi command mode behaves like this:
* `ab|c` => `ab|`  (expected)
* `ab|` => `a|b`  (moves left but doesn't delete anything, wth??)
* `|` => ??? (it causes readline to exit when the line is empty, wth??)

Note that not only does `em_delete` not work properly in vi command mode (which shouldn't be too surprising since we're not in emacs mode), but also `ed_delete_next_char` doesn't work properly in vi command mode either (and perhaps that's the only real reason why it's not allowed right now).

Finally, take note of the fact that both `set_default_key_bindings_comprehensive_list` in ansi.rb as well as `set_default_key_bindings` in windows.rb map the delete key to `key_delete`, not `em_delete`.

I do think `key_delete` is not behaving the best way, because I would totally argue for not moving left when deleting the last character, and I would also argue for making it actually work in vi command mode too.  But I think fixing `key_delete` just needs to be written up as its own issue.